### ISSUE TITLE
ClassFormatException: Constant pool at index 0 is null

### ISSUE THREAD

**Issue by barney2k7** [spotbugs/spotbugs#60]
The problem described in findbugsproject/findbugs#141 also affects spotbugs, I just verified it with a current snapshot. The analysis of the following code fails:
```
import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.util.Properties;
import java.util.stream.Stream;

public class MyClass {
  
  public static void create(URL url) {
    try (InputStream in = url.openStream()) {
      Properties p1 = new Properties();
      p1.load(in);
    } catch (IOException e) {
    }
  }
  
  public Stream<String> keys() {
    return Stream.<Properties>of().flatMap(p -> p.stringPropertyNames().stream());
  }

}
```

Stacktrace:
```
The following errors occurred during analysis:
  Exception analyzing MyClass using detector edu.umd.cs.findbugs.detect.FindUnsatisfiedObligation
    org.apache.bcel.classfile.ClassFormatException: Constant pool at index 0 is null.
      At org.apache.bcel.classfile.ConstantPool.getConstant(ConstantPool.java:258)
      At org.apache.bcel.classfile.ConstantPool.getConstantString(ConstantPool.java:293)
      At org.apache.bcel.generic.FieldOrMethod.getReferenceType(FieldOrMethod.java:116)
      At edu.umd.cs.findbugs.ba.obl.InstructionActionCache.getActions(InstructionActionCache.java:107)
      At edu.umd.cs.findbugs.ba.obl.ObligationAnalysis.transferInstruction(ObligationAnalysis.java:139)
      At edu.umd.cs.findbugs.ba.obl.ObligationAnalysis.transferInstruction(ObligationAnalysis.java:68)
      At edu.umd.cs.findbugs.ba.AbstractDataflowAnalysis.transfer(AbstractDataflowAnalysis.java:135)
      At edu.umd.cs.findbugs.ba.obl.ObligationAnalysis.transfer(ObligationAnalysis.java:160)
      At edu.umd.cs.findbugs.ba.obl.ObligationAnalysis.transfer(ObligationAnalysis.java:68)
      At edu.umd.cs.findbugs.ba.Dataflow.execute(Dataflow.java:376)
      At edu.umd.cs.findbugs.classfile.engine.bcel.ObligationDataflowFactory.analyze(ObligationDataflowFactory.java:75)
      At edu.umd.cs.findbugs.classfile.engine.bcel.ObligationDataflowFactory.analyze(ObligationDataflowFactory.java:45)
      At edu.umd.cs.findbugs.classfile.impl.AnalysisCache.analyzeMethod(AnalysisCache.java:369)
      At edu.umd.cs.findbugs.classfile.impl.AnalysisCache.getMethodAnalysis(AnalysisCache.java:322)
      At edu.umd.cs.findbugs.detect.FindUnsatisfiedObligation$MethodChecker.analyzeMethod(FindUnsatisfiedObligation.java:240)
      At edu.umd.cs.findbugs.detect.FindUnsatisfiedObligation.visitMethodCFG(FindUnsatisfiedObligation.java:158)
      At edu.umd.cs.findbugs.bcel.CFGDetector.visitClass(CFGDetector.java:93)
      At edu.umd.cs.findbugs.detect.FindUnsatisfiedObligation.visitClass(FindUnsatisfiedObligation.java:138)
      At edu.umd.cs.findbugs.FindBugs2.analyzeApplication(FindBugs2.java:1089)
      At edu.umd.cs.findbugs.FindBugs2.execute(FindBugs2.java:283)
      At edu.umd.cs.findbugs.gui2.BugLoader.doAnalysis(BugLoader.java:100)
      At edu.umd.cs.findbugs.gui2.AnalyzingDialog$AnalysisThread.run(AnalyzingDialog.java:274)

```

Thanks!

**Comment by KengoTODA** [comment]
I will try to reproduce this in `issue-20` branch.

**Comment by KengoTODA** [comment]
I'm still testing #61, which is necessary to test your class with our integration test. Here is current patch:

```patch
diff --git a/findbugs/src/test/java/edu/umd/cs/findbugs/detect/FindUnsatisfiedObligationTest.java b/findbugs/src/test/java/edu/umd/cs/findbugs/detect/FindUnsatisfiedObligationTest.java
new file mode 100644
index 0000000..a767df9
--- /dev/null
+++ b/findbugs/src/test/java/edu/umd/cs/findbugs/detect/FindUnsatisfiedObligationTest.java
@@ -0,0 +1,22 @@
+package edu.umd.cs.findbugs.detect;
+
+import static org.junit.Assert.assertThat;
+
+import static org.hamcrest.core.Is.is;
+
+import static org.hamcrest.collection.IsEmptyIterable.*;
+import org.junit.Test;
+
+import edu.umd.cs.findbugs.AbstractIntegrationTest;
+
+public class FindUnsatisfiedObligationTest extends AbstractIntegrationTest {
+    /**
+     * @see <a href="https://github.com/spotbugs/spotbugs/issues/60">GitHub
+     *      issue</a>
+     */
+    @Test
+    public void testIssue60() {
+        performAnalysis("Issue60.class");
+        assertThat(getBugCollection(), is(emptyIterable()));
+    }
+}
diff --git a/findbugsTestCases/src/java/Issue60.java b/findbugsTestCases/src/java/Issue60.java
new file mode 100644
index 0000000..b79f1b8
--- /dev/null
+++ b/findbugsTestCases/src/java/Issue60.java
@@ -0,0 +1,22 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.Properties;
+import java.util.stream.Stream;
+
+public class Issue60 {
+
+    public static void create(URL url) {
+        try (InputStream in = url.openStream()) {
+            Properties p1 = new Properties();
+            p1.load(in);
+        } catch (IOException e) {
+        }
+    }
+
+    public Stream<String> keys() {
+        return Stream.<Properties> of()
+                .flatMap(p -> p.stringPropertyNames().stream());
+    }
+
+}
```

**Comment by KengoTODA** [comment]
By #69, I confirmed that current `master` branch has no problem. @barney2k7 please have a try.

**Comment by jsotuyod** [comment]
@KengoTODA the exception does occur, but is managed and simply logged to stderr without the test failing.

**Comment by KengoTODA** [comment]
@jsotuyod Oh... I also confirmed this behaviour.

**Comment by jsotuyod** [comment]
Running `javap` on the compiled source we can see the offending code:

```
  public java.util.stream.Stream<java.lang.String> keys();
    descriptor: ()Ljava/util/stream/Stream;
    flags: ACC_PUBLIC
    Code:
      stack=2, locals=1, args_size=1
         0: iconst_0
         1: anewarray     #3                  // class java/util/Properties
         4: invokestatic  #10                 // InterfaceMethod java/util/stream/Stream.of:([Ljava/lang/Object;)Ljava/util/stream/Stream;
         7: invokedynamic #11,  0             // InvokeDynamic #0:apply:()Ljava/util/function/Function;
```

And in the constant pool we can see:

```
Constant pool:
    #1 = Methodref          #16.#46       // java/lang/Object."<init>":()V
    #2 = Methodref          #47.#48       // java/net/URL.openStream:()Ljava/io/InputStream;
    #3 = Class              #49           // java/util/Properties
    #4 = Methodref          #3.#46        // java/util/Properties."<init>":()V
    #5 = Methodref          #3.#50        // java/util/Properties.load:(Ljava/io/InputStream;)V
    #6 = Methodref          #51.#52       // java/io/InputStream.close:()V
    #7 = Class              #53           // java/lang/Throwable
    #8 = Methodref          #7.#54        // java/lang/Throwable.addSuppressed:(Ljava/lang/Throwable;)V
    #9 = Class              #55           // java/io/IOException
   #10 = InterfaceMethodref #56.#57       // java/util/stream/Stream.of:([Ljava/lang/Object;)Ljava/util/stream/Stream;
   #11 = InvokeDynamic      #0:#63        // #0:apply:()Ljava/util/function/Function;
```

This happens because the `invokedynamic` is acting on the method handle, which of course, has no class name.

This is a bug on BCEL, that should override `getReferenceType` on `INVOKEDYNAMIC` to at least not throw a `ClassFormatException`. What do you think @mebigfatguy ?

In the meantime, we may rewrite the detector to handle `INVOKEDYNAMIC` separately.

**Comment by mebigfatguy** [comment]
yeah, should add a bug here http://issues.apache.org/jira/browse/BCEL

**Comment by mebigfatguy** [comment]
i suppose it should return Object?

**Comment by jsotuyod** [comment]
You are amazing! Thanks [for the quick fix](https://github.com/apache/commons-bcel/commit/28d51dace3fbbf8f0bd71f749e5290685b0c5df2)

I guess for the time being we can wait for a BCEL 6.1 release / use a snapshot until we reach the point of releasing Spotbugs 3.1.0

**Comment by KengoTODA** [comment]
This problem should be solved by #69. I will close this issue.
